CsrfCounterMeasure: accept Sec-Fetch-Site header, if available, instead of token

Al2Klimov opened the pull request:
This is especially useful if the session and token change suddenly, e.g. due to mod_auth_openidc.

refs https://github.com/Icinga/icingaweb2/issues/5224

Al2Klimov committed:
CsrfCounterMeasure: accept Sec-Fetch-Site header, if available, instead of token

This is especially useful if the session and token change suddenly, e.g. due to mod_auth_openidc.

TheSyscall commented on src/Common/CsrfCounterMeasure.php:
```diff
@@ -108,4 +112,13 @@ protected function addCsrfCounterMeasure(?string $uniqueId = null): void
         $this->addElement($this->createCsrfCounterMeasure($uniqueId ?? $this->csrfCounterMeasureId));
     }
+
+    protected function requestIsSafe(): bool
```
Missing docstring

TheSyscall commented on src/Common/CsrfCounterMeasure.php:
```diff
@@ -72,6 +72,10 @@ protected function createCsrfCounterMeasure($uniqueId)
+        if ($this->requestIsSafe()) {
+            unset($options['validators']);
+        }
+
```
Why not include this check as part of the validator?
It should only run when the form is actually sent. `addCsrfCounterMeasure` and `createCsrfCounterMeasure` are called from within `assemble`, so with any request type.
Adding the full CSRF token with validator in every request unconditionally makes sense for consistency, even if we end up ignoring its value later.

TheSyscall commented on src/Common/CsrfCounterMeasure.php:
```diff
@@ -98,7 +102,7 @@ public function hasValue(): bool
     protected function addCsrfCounterMeasure(?string $uniqueId = null): void
     {
-        if (! $this->csrfCounterMeasureEnabled) {
+        if (! $this->csrfCounterMeasureEnabled || $this->requestIsSafe()) {
```
See above

Al2Klimov replied:
That's on purpose! I want to make setting the token optional.

nilmerg reviewed (changes requested):
This misses an important point, to me. Either that or it's not documented/mentioned at all.

The token was always only a mitigation for CSRF, hence the name CSRFToken. It does **not** authenticate/authorize. So by supporting this request header, we fully trust the browser. But why is there still a token in case the header value is `same-site` or `cross-site`? Why doesn't it simply fail?

If the header is set the browser supports it and sets it reliably with every request. Or is this not guaranteed? With the header set, the initial response can decide whether a token is required or not and any subsequent request either includes the header again or a token.

There's just no reality I see that a form is being submitted *legitimately* if the header says it's not `same-origin`.

nilmerg commented on src/Common/CsrfCounterMeasure.php:
```diff
@@ -69,8 +74,12 @@ protected function createCsrfCounterMeasure($uniqueId)
+            }];
+        } else if (! $requestIsSafe) {
+            $options['validators'] = ['Callback' => function () {
+                throw new Error('Rejecting cross-site request');
```
`createCsrfCounterMeasure` is deprecated. There's no reason to return an element in case the request isn't considered safe. Neither does `addCsrfCounterMeasure` and throws immediately.

So instead of refactoring the entire method, let it early return or throw as well.

Al2Klimov replied:
> instead of refactoring the entire method, let it early return

Would this be acceptable for a safe request?

```php
return new HiddenElement('CSRFToken', ['ignore' => true]);
```

nilmerg replied:
No, by looking at users of the trait I noticed one that expects this to return an element with a validator set that verifies a manually set token. Quite hacky, yes, but nonetheless, the method `create…` must still return an element and that should have at least a validator that succeeds.

Al2Klimov replied:
Better?

```php
return new HiddenElement('CSRFToken', [
    'ignore'     => true,
    'validators' => ['Callback' => function () {
        return true;
    }]
]);
```

nilmerg replied:
Obviously? Now start committing please…

nilmerg reviewed (changes requested):
* Please also adjust the accompanying test

Another important point to clarify:

The rationale for solely relying on this header instead of still requiring a token is that we only care about real browsers. Clients such as cURL will be able to spoof this header but in reality, they never needed to provide a token in the first place as if the route in question supports direct access like this, transmitting it as `Content-Type: application/json` does not require it anyway. Any route not supporting this still requires the user to authenticate and this then means the client needs to posess a session cookie and is able to also forge a CSRF token on its own.

TheSyscall commented on src/Common/CsrfCounterMeasure.php:
```diff
@@ -102,10 +117,39 @@ protected function addCsrfCounterMeasure(?string $uniqueId = null): void
+
+            return;
+        }
+
```
Why is this checked twice? Once in here and then again in `createCsrfCounterMeasure`.

As far as I can see skips only the check for the `csrfCounterMeasureId`, which should already be present for the form to be assembled anyway.

Al2Klimov replied:
That's on purpose to allow omitting setCsrfCounterMeasureId(). After all, we also omit token generation now for new browsers.

TheSyscall replied:
This would lead to users on older browsers causing Errors/LogicExceptions #386 on the server, because `setCsrfCounterMeasureId()` was never called in the expectation that every request to this form can only come from new browsers.

We essentially set up Csrf protection with the expectation to never generate a CSRF-Token.

Al2Klimov replied:
I made the message more clear.

TheSyscall replied:
I'm still not quite happy with this code, it hides a potential implementation error behind an expectation on the users' browser.

@nilmerg What do you say to this?

nilmerg replied:
@TheSyscall is right, I say. The error below signals a programming error (and should have used LogicException tbh) and in case said programmer uses a modern browser, his/her/its user with a legacy browser runs into it.

Al2Klimov replied:
Done.

TheSyscall commented on src/Common/CsrfCounterMeasure.php:
The request is safe behavior should be documented.

TheSyscall commented on src/Common/CsrfCounterMeasure.php:
The request is safe behavior should be documented.

TheSyscall commented on src/Common/CsrfCounterMeasure.php:
```diff
@@ -102,10 +117,39 @@ protected function addCsrfCounterMeasure(?string $uniqueId = null): void
+     * Get whether the request is safe from CSRF based on the Sec-Fetch-Site header
+     *
+     * Returns true if the header indicates a same-origin request or a user-originated operation (e.g. typing a URL),
+     * both of which cannot be forged by a cross-site attacker. Returns null if the header is absent,
```
Maybe add clarification that it can't be forged with JavaScript by a browser

Al2Klimov replied:
Well... that's what I say here?

Al2Klimov committed:
CsrfCounterMeasure: accept Sec-Fetch-Site header, if available, instead of token

This is especially useful if the session and token change suddenly, e.g. due to mod_auth_openidc.

TheSyscall reviewed (approved):
👍

nilmerg reviewed (approved):
I verified how a browser handles `Sec-Fetch-Site: none` in case of service workers, as I was curios whether it is okay that this change concludes that the request is safe in this case. For this, I let a service-worker control every network request and monitored which value is sent. The observation I made was that until the request is transmitted using the worker, no Sec-Fetch-Site header is part of the request. Only when the worker transmits it, the browser detects this as programmatic fetch, the value is set to `same-origin` or `cross-origin`.

So my paranoia-me is satisfied.
